### ISSUE TITLE
Negative LIMIT specification.

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#28913]
**Use case**

`LIMIT -10`
\- return 10 last records (regardless of total number of records).

Can be combined with negative offset (but not with positive offset).

**Describe the solution you'd like**

Just implement it.

**Describe alternatives you've considered**

`SELECT * (SELECT ... ORDER BY x DESC LIMIT 10) ORDER BY x ASC`

But it's more cumbersome.

**Comment by Joeywzr** [comment]
Could you assign this task to me? It has been 2 years since my last pr. I think I need a "warm-up", hah.

**Comment by alexey-milovidov** [comment]
Sure! It's yours :)

**Comment by alexey-milovidov** [comment]
LimitTransform.h / LimitTransform.cpp

Keep the queue of last blocks that have at least negative limit + negative offset rows. When stream ends, cut these blocks accordingly.

**Comment by Joeywzr** [comment]
got it

**Comment by Joeywzr** [comment]
@alexey-milovidov If the limit entered by the user is negative and the offset is positive, do we need to throw an exception or just treat the offset as negative？

**Comment by alexey-milovidov** [comment]
Throw exception.

**Comment by Joeywzr** [comment]
@alexey-milovidov I've finished the part except `order by xxx`. It seems that I need to modify a lot of files like `PartialSortingTransform`, `MergeSortingTransform`, `ColumnXXXX` to fit it.  I wonder if there is something wrong with my understanding.

**Comment by alexey-milovidov** [comment]
If limit is negative, we can omit the partial sorting optimization (as if there is no limit) - so there is no need to modify these places.

**Comment by Joeywzr** [comment]
@alexey-milovidov Thanks! That helps me a lot.

**Comment by Joeywzr** [comment]
#29496

**Comment by Joeywzr** [comment]
@alexey-milovidov pls add a `can be test` label.

**Comment by taofengliu** [comment]
Since there has been no progress since April, can I help to finish this?

**Comment by Joeywzr** [comment]
> 

Sure.

**Comment by taofengliu** [comment]
Starting work on it.

**Comment by CheSema** [comment]
In that issue has been given definition for negative limit. What is definition for negative offset? Why do limit and offset have to be the same sign?

**Comment by alexey-milovidov** [comment]
A motivating case for negative limit:

`SELECT str FROM (SELECT event_time_microseconds, format('{0} [ {1} ] ({2}) <{3}> {4}: {5}', event_time_microseconds, thread_id, query_id, level, logger_name, message) AS str FROM system.text_log ORDER BY event_time DESC, event_time_microseconds DESC LIMIT 100) ORDER BY event_time_microseconds FORMAT TSVRaw`

**Comment by nihalzp** [comment]
@alexey-milovidov I will be happy to work on this feature. The existing PRs appear to have been inactive for over a year.

**Comment by alexey-milovidov** [comment]
@nihalzp, yes, this will be very appreciated! This feature is highly demanded.

**Comment by 0xgouda** [comment]
Picking this up.

**Comment by nihalzp** [comment]
> Picking this up.

@0xgouda Apologies. I started working on this one week earlier and should be done soon.

**Comment by 0xgouda** [comment]
@nihalzp No problem, I hope you do the best!
